### PR TITLE
Support sampling (for V5)

### PR DESCRIPTION
**Pull request by kiryk**
The change adds support for value sampling, i.e. access to values of expressions at the beginning of each time slot.

This is done with `$sampled` function except for concurrent assertions where sampling is done by default.
Before the change, the function was deleted during V3Assert pass when encountered in AST.

The change makes `$past`, `$fell`, `$rose`, `$stable` and `$changed` refer to sampled values, as expected by IEEE.
The change can simplify implementation of explicit clocking for these functions, but such functionality is not added here.

Sampling is implemented in three stages:
1. On parsing stage concurrent asserts are created with `$sampled` around their test expressions.
2. On V3Assert stage each variable reference inside a sampled expression is wrapped into a call to `$sampled`, all other uses of `$sampled` are removed.
3. On V3Clock stage every call to `$sampled` is replaced by their only subexpression, and for every variable in the subexpression:
- a helper variable holding the sampled value is created,
- an assignment to the variable is added to the `eval` to occur at the beginning of every simulation step,
- the original variable reference is replaced by a reference to the helper variable.

The code that creates variables used for keeping sampled values, is placed in V3Clock pass. The reason for that is that it's the earliest pass that runs after the `_eval` function is created and can be accessed for modification. V3Clock also runs before any of the final optimizations are performed.

It is undefined what should happen if a sampled expression has side effects. Expressions like `$sampled(a++)` and `$sampled(f(a))` (where `f` has side effects) behave very differently in various SystemVerilog simulators. This PR makes Verilator produce an `E_UNSUPPORTED` warning in the first case, but the second case is harder to detect, and thus ignored.

### CODE REVIEW

File: test_regress/t/t_assert_past.pl [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+# Copyright 2022 by Antmicro Ltd. This program is free software; you
+# can redistribute it and/or modify it under the terms of either the GNU
+# Lesser General Public License Version 3 or the Perl Artistic License
+# Version 2.0.
+# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+
+scenarios(simulator => 1);
+
+compile(
+    verilator_flags2 => ['--assert --cc'],
```
Comment (kbieganski) [inline test_regress/t/t_assert_past.pl:14]: I think `--cc` isn't needed here. (Same in other test cases)

File: src/V3Clock.cpp [hunk 2]
```diff
@@ -151,12 +175,58 @@ class ClockVisitor final : public VNVisitor {
...
 
+    // Create sampled values
+    virtual void visit(AstNodeModule* nodep) override {
+        VL_RESTORER(m_modp);
+        {
+            m_modp = nodep;
+            iterateChildren(nodep);
+        }
+    }
+    virtual void visit(AstTopScope* nodep) override {
```
Comment (kbieganski) [inline src/V3Clock.cpp:186]: This function is unnecessary. The scope underneath will be visited anyway.

File: src/V3Clock.cpp [hunk 3]
```diff
@@ -151,12 +175,58 @@ class ClockVisitor final : public VNVisitor {
         clearLastSen();
     }
 
+    // Create sampled values
+    virtual void visit(AstNodeModule* nodep) override {
```
Comment (kbieganski) [inline src/V3Clock.cpp:179]: We don't need to save the module. We can get it from `m_scopep->modp()`.

File: test_regress/t/t_sampled_expr.v [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+endmodule
+
+module Test1(
+   clk, a, b
+   );
+
+   input clk;
+   input [3:0] a, b;
+
+   assert property (@(posedge clk) $sampled(a == b) == ($sampled(a) == $sampled(b)));
```
Comment (kbieganski) [inline test_regress/t/t_sampled_expr.v:43]: Don't we need some sort of `else $stop;` here? Do asserts fail on their own? (Same in other test cases)

File: src/V3Clock.cpp [hunk 5]
```diff
@@ -151,12 +175,58 @@ class ClockVisitor final : public VNVisitor {
...
+            nodep->replaceWith(nodep->exprp()->unlinkFrBack());
+            VL_DO_DANGLING(pushDeletep(nodep), nodep);
+        }
+    }
+    virtual void visit(AstVarRef* nodep) override {
+        iterateChildren(nodep);
+        if (m_inSampled && !nodep->user1SetOnce()) {
+            AstVarScope* const varscp = nodep->varScopep();
+            AstVarScope* const lastscp = createSampledVar(varscp);
+            AstNode* const newp = new AstVarRef(nodep->fileline(), lastscp, VAccess::READ);
```
Comment (kbieganski) [inline src/V3Clock.cpp:214]:
```suggestion
            AstNode* const newp = new AstVarRef{nodep->fileline(), lastscp, VAccess::READ};
```
Also, is it possible for `nodep` to be `WRITE`? If it doesn't error out earlier, add a `v3error` here. If it does, add an assert here.

File: src/V3Clock.cpp [hunk 6]
```diff
@@ -87,6 +91,26 @@ class ClockVisitor final : public VNVisitor {
...
+            = new AstVar(vscp->fileline(), VVarType::MODULETEMP, newvarname, varp->dtypep());
+        newvarp->noReset(true);  // Reset by below assign
+        m_modp->addStmtp(newvarp);
+        AstVarScope* const newvscp = new AstVarScope(vscp->fileline(), m_scopep, newvarp);
+        vscp->user1p(newvscp);
+        m_scopep->addVarp(newvscp);
+        // At the top of _eval, assign them
+        AstAssign* const finalp = new AstAssign(
+            vscp->fileline(), new AstVarRef(vscp->fileline(), newvscp, VAccess::WRITE),
+            new AstVarRef(vscp->fileline(), vscp, VAccess::READ));
```
Comment (kbieganski) [inline src/V3Clock.cpp:109]:
```suggestion
            = string("__Vsampled__") + vscp->scopep()->nameDotless() + "__" + varp->name();
        FileLine* const flp = vscp->fileline();
        AstVar* const newvarp
            = new AstVar{flp, VVarType::MODULETEMP, newvarname, varp->dtypep()};
        newvarp->noReset(true);  // Reset by below assign
        m_modp->addStmtp(newvarp);
        AstVarScope* const newvscp = new AstVarScope{flp, m_scopep, newvarp};
        vscp->user1p(newvscp);
        m_scopep->addVarp(newvscp);
        // At the top of _eval, assign them
        AstAssign* const finalp = new AstAssign{
            flp, new AstVarRef{flp, newvscp, VAccess::WRITE},
            new AstVarRef{flp, vscp, VAccess::READ}};
```

File: src/V3Clock.cpp [hunk 7]
```diff
@@ -72,8 +72,12 @@ class ConvertWriteRefsToRead final : public VNVisitor {
 class ClockVisitor final : public VNVisitor {
 private:
     // STATE
+    AstCFunc* m_evalp = nullptr;  // Pointer to _eval
```
Comment (kbieganski) [inline src/V3Clock.cpp:75]:
```suggestion
    AstCFunc* m_evalp = nullptr;  // The '_eval' function
```

**Review by kbieganski — CHANGES_REQUESTED** [review]

File: test_regress/t/t_sampled_expr.v (hunk 4 above)
Comment (kiryk) [reply test_regress/t/t_sampled_expr.v:43]: We don't, assertions restrain from failing automatically only when they have an else, otherwise it's a default behavior.

File: test_regress/t/t_assert_past.pl (hunk 1 above)
Comment (kiryk) [reply test_regress/t/t_assert_past.pl]: I deleted these.

File: src/V3Clock.cpp (hunk 2 above)
Comment (kiryk) [reply src/V3Clock.cpp]: Thanks, I wasn't sure about that one. Deleted.

File: src/V3Clock.cpp (hunk 3 above)
Comment (kiryk) [reply src/V3Clock.cpp]: Done.

File: src/V3Clock.cpp (hunk 5 above)
Comment (kiryk) [reply src/V3Clock.cpp]: I errored it out, but it's an interesting case. The grammar of Verilator only allows expressions there, so no value can be modified except for incrementations and decrementations.

Since values accessed in asserts are sampled, it seems unclear what should happen when one of these two is used.

File: src/V3Clock.cpp (hunk 6 above)
Comment (kiryk) [reply src/V3Clock.cpp]: I've applied these.

File: src/V3Clock.cpp (hunk 7 above)
Comment (kiryk) [reply src/V3Clock.cpp]: Done.

File: src/V3Clock.cpp (hunk 5 above)
Comment (kbieganski) [reply src/V3Clock.cpp]: Nice catch. So does the LRM allow incrementations/decrementations there? If not, we should make it clear it's illegal and add a test. If it does, and handling that case is difficult, we should warn that it's unsupported. This probably won't be caught at this stage, as various transformations along the way would make it into a normal expression.

What about output function arguments and non-pure functions? Seems like there's huge potential for bugs. Maybe we should also treat function calls in `$sampled` as unsupported.

Please add tests for these edge cases (unless they turn out to be trivial).

Comment (kiryk) [reply src/V3Clock.cpp]: Okay, so I made a few changes:
* writes to variables in sampled areas stay errored out as unsupported,
* pure functions work well now,
* function side effects have undefined behavior,
* sampling in concurrent asserts is injected on parsing stage.

I also made a few clean ups.

To sum up:
Our problems came from the fact that we want $sampled to be eventually applied to variables only, but optimizations can break it in two ways:
1. Moving a sampled variable from inside a sampled expression to another place (like it was with function arguments).
2. Wrapping a $sampled variable into something extra, thus breaking the desired state where only variables are sampled.

Now we're ready for 1. because at an early stage (V3Assert) we simplify sampled expressions to ones that only have sampled variables. So in case an optimization moves a part of an expression somewhere else, all variables go with their $sampled.

We're also ready for 2. because at a late stage (V3Clock) where calls to $sampled are replaced by helper variables, we don't care if the argument is actually a variable, we just recursively look for every variable inside the call, and replace it with a helper.

File: src/V3Assert.cpp [hunk 8]
```diff
@@ -353,10 +360,40 @@ class AssertVisitor final : public VNVisitor {
...
+            iterateChildren(nodep);
+        }
         nodep->replaceWith(nodep->exprp()->unlinkFrBack());
         VL_DO_DANGLING(pushDeletep(nodep), nodep);
     }
+    virtual void visit(AstVarRef* nodep) override {
+        iterateChildren(nodep);
+        if (!m_inSampled) return;
+        if (!nodep->access().isReadOnly()) {
+            nodep->v3warn(E_UNSUPPORTED,
```
Comment (kbieganski) [inline src/V3Assert.cpp:379]: Is it actually possible to hit this branch? If so, add a test. Else replace with an assert.

File: src/V3Clock.cpp (hunk 5 above)
Comment (kbieganski) [reply src/V3Clock.cpp]: Sounds good. See my other remaining comment.

File: src/V3Assert.cpp (hunk 8 above)
Comment (kiryk) [reply src/V3Assert.cpp:379]: It is possible. I've added the test.

File: test_regress/t/t_sampled_expr_bad.pl [hunk 9]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env perl
+if (!$::Driver) { use FindBin; exec("$FindBin::Bin/bootstrap.pl", @ARGV, $0); die; }
+# DESCRIPTION: Verilator: Verilog Test driver/expect definition
+#
+# Copyright 2019 by Wilson Snyder. This program is free software; you
```
Comment (kbieganski) [inline test_regress/t/t_sampled_expr_bad.pl:5]: Please fix the copyright, rename this test to `t_sampled_expr_unsup.pl`, and add an output comparison. See other tests, look for `expect_filename => $Self->{golden_filename}`. You need a `t_sampled_expr_unsup.out` file for it, but don't edit it manually, just `touch` it and then run the test with `--golden`.

File: test_regress/t/t_sampled_expr_bad.v [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+   input clk;
+
+   integer cyc;
+
+   Test1 t1(clk);
+
+   always @(posedge clk) begin
+      cyc <= cyc + 1;
+
+      $display("$time = %0d, cyc = %0d", $time, $past($time), cyc);
```
Comment (kbieganski) [inline test_regress/t/t_sampled_expr_bad.v:20]: We preferably shouldn't use `$past` in a `$sampled` test. Let's remove this `$display`, it's not needed for this test (we never run it).

**Review by kbieganski — CHANGES_REQUESTED** [review]
Just some final small stuff.

File: test_regress/t/t_sampled_expr_bad.pl (hunk 9 above)
Comment (kiryk) [reply test_regress/t/t_sampled_expr_bad.pl]: Thanks for detailed instruction. Done.

File: test_regress/t/t_sampled_expr_bad.v (hunk 10 above)
Comment (kiryk) [reply test_regress/t/t_sampled_expr_bad.v]: Done.

File: src/V3Assert.cpp [hunk 11]
```diff
@@ -353,10 +360,40 @@ class AssertVisitor final : public VNVisitor {
...
+                          "Unsupported: Assignment to variable inside sampled expression");
```
Comment (kbieganski) [inline src/V3Assert.cpp:380]: Let's reword this a bit.
```suggestion
                          "Unsupported: Write to variable in sampled expression");
```

File: test_regress/t/t_sampled_expr_unsup.out [hunk 12]
```diff
@@ -0,0 +1,11 @@
+%Error-UNSUPPORTED: t/t_sampled_expr_unsup.v:36:36: Unsupported: Assignment to variable inside sampled expression
+                                                  : ... In instance t.t1
+   36 |    assert property (@(posedge clk) a++ >= 0);
+      |                                    ^
+                    ... For error description see https://verilator.org/warn/UNSUPPORTED?v=latest
+%Warning-UNSIGNED: t/t_sampled_expr_unsup.v:36:40: Comparison is constant due to unsigned arithmetic
```
Comment (kbieganski) [inline test_regress/t/t_sampled_expr_unsup.out:6]: Please get rid of this warning. Preferably by using a different type/comparison, but I'll accept `-Wno-UNSIGNED`.

**Review by kbieganski — CHANGES_REQUESTED** [review]
Ok, these are really final nitpicks 😄

File: test_regress/t/t_sampled_expr_unsup.out (hunk 12 above)
Comment (kiryk) [reply test_regress/t/t_sampled_expr_unsup.out]: I opt for `-Wno-UNSIGNED` because I intended the test to be tautological for it not to fail for any other reason than just "write in sampled".

I updated the test.

File: src/V3Assert.cpp (hunk 11 above)
Comment (kiryk) [reply src/V3Assert.cpp]: Done.

Comment (kbieganski) [reply src/V3Assert.cpp]: Update the `*.out` file as well.

Comment (kiryk) [reply src/V3Assert.cpp]: I've pushed it meanwhile.

**Review by kbieganski — APPROVED** [review]
Look great! Once the CI passes, open a PR to mainline.
